File: app/src/main/java/com/weatherapp/service/CurrentLocationManager.java
package com.weatherapp.service;

import android.annotation.SuppressLint;
import android.content.Context;
import android.location.Location;
import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import com.google.android.gms.location.*;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;

public class CurrentLocationManager extends LiveData<Location> {

    private static CurrentLocationManager instance;
    private FusedLocationProviderClient mFusedLocationClient;
    private LocationRequest mLocationRequest;

    public static CurrentLocationManager getInstance(Context appContext) {
        if (instance == null) {
            instance = new CurrentLocationManager(appContext);
        }
        return instance;
    }

    @SuppressLint("MissingPermission")
    private CurrentLocationManager(final Context appContext) {
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(appContext);
        createLocationRequest();
        mFusedLocationClient.getLastLocation().addOnCompleteListener(new OnCompleteListener<Location>() {
            @SuppressLint("MissingPermission")
            @Override
            public void onComplete(@NonNull Task<Location> task) {
                if(task.isSuccessful() && task.getResult() != null) {
                    setValue(task.getResult());
                } else {
                    mFusedLocationClient.requestLocationUpdates(mLocationRequest, mLocationCallback, null);
                }
            }
        });

    }

    private void createLocationRequest() {
        mLocationRequest = new LocationRequest();
        mLocationRequest.setInterval(10000);
        mLocationRequest.setFastestInterval(5000);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
    }

    LocationCallback mLocationCallback = new LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            for (Location location : locationResult.getLocations()) {
                if (location != null)
                    setValue(location);
            }
        }
    };

    @Override
    protected void onInactive() {
        super.onInactive();
        if (mLocationCallback != null)
            mFusedLocationClient.removeLocationUpdates(mLocationCallback);
    }

}

File: app/src/main/java/com/weatherapp/service/CurrentLocationManager.kt
package com.weatherapp.service

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import androidx.lifecycle.LiveData
import com.google.android.gms.location.*
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task

class CurrentLocationManager private constructor(appContext: Context) : LiveData<Location>() {
    private val mFusedLocationClient: FusedLocationProviderClient
    private var mLocationRequest: LocationRequest? = null

    internal var mLocationCallback: LocationCallback? = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            for (location in locationResult.getLocations()) {
                if (location != null)
                    setValue(location)
            }
        }
    }

    init {
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(appContext)
        createLocationRequest()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (appContext.checkSelfPermission(android.Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                initLocation()
            }
        } else {
            initLocation()
        }

    }


    @SuppressLint("MissingPermission")
    fun initLocation() {
        mFusedLocationClient.getLastLocation().addOnCompleteListener(object : OnCompleteListener<Location> {
            @SuppressLint("MissingPermission")
            override fun onComplete(task: Task<Location>) {
                if (task.isSuccessful() && task.getResult() != null) {
                    setValue(task.getResult())
                } else {
                    mFusedLocationClient.requestLocationUpdates(mLocationRequest, mLocationCallback, null)
                }
            }
        })

    }

    private fun createLocationRequest() {
        mLocationRequest =
            LocationRequest()
        mLocationRequest?.interval = 10000
        mLocationRequest?.fastestInterval = 5000
        mLocationRequest?.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
    }

    protected override fun onInactive() {
        super.onInactive()
        if (mLocationCallback != null)
            mFusedLocationClient.removeLocationUpdates(mLocationCallback)
    }

    companion object {

        private var instance: CurrentLocationManager? = null

        fun getInstance(appContext: Context): CurrentLocationManager {
            if (instance == null) {
                instance = CurrentLocationManager(appContext)
            }
            return instance!!
        }
    }

}
